rbs/CommentsAssociator.cc

```diff
@@ -16,22 +17,117 @@ const std::string_view CommentsAssociator::ANNOTATION_PREFIX = "# @";
+    regex heredoc_pattern("\\s*=?\\s*<<(-|~)[^,\\s\\n#]+(,\\s*<<(-|~)[^,\\s\\n#]+)*");
+    smatch match;
+    if (regex_search(source, match, heredoc_pattern)) {
+        return fromPos + match.length();
+    }
+    return UINT32_MAX;
 }
+}; // namespace
+
+uint32_t CommentsAssociator::heredocPos(core::Context ctx, core::LocOffsets assignLoc, parser::Node *node) {
+    if (node == nullptr) {
+        return UINT32_MAX;
+    }
+
+    uint32_t result = UINT32_MAX;
+    typecase(
+        node,
+        [&](parser::String *lit) {
+            // For some reason, heredoc strings are parser differently if they contain a single line or more.
+            //
+            // Single line heredocs do not contain the `<<-` or `<<~` markers inside their location.
+            //
+            // For example, this heredoc:
+            //
+            //     <<~MSG
+            //       foo
+            //     MSG
+
+            // has the `<<-` or `<<~` markers **outside** its location.
+            //
+            // While this heredoc:
+            //
+            //     <<~MSG
+            //       foo
+            //     MSG
+            //
+            // has the `<<-` or `<<~` markers **inside** its location.
+
+            auto lineStart = core::Loc::pos2Detail(ctx.file.data(ctx), lit->loc.beginLoc).line;
+            auto lineEnd = core::Loc::pos2Detail(ctx.file.data(ctx), lit->loc.endLoc).line;
+
+            if (lineEnd - lineStart <= 1) {
+                // Single line heredoc, we look for the heredoc marker outside, ie. between the assign `=` sign
+                // and the begining of the string.
+                result = hasHeredocMarker(ctx, assignLoc.endPos(), lit->loc.beginPos());
+            } else {
+                // Multi-line heredoc, we look for the heredoc marker inside the string itself.
+                result = hasHeredocMarker(ctx, lit->loc.beginPos(), lit->loc.endPos());
+            }
+        },
+        [&](parser::DString *lit) { result = hasHeredocMarker(ctx, lit->loc.beginPos(), lit->loc.endPos()); },
+        [&](parser::Array *arr) {
+            for (auto &elem : arr->elts) {
+                result = heredocPos(ctx, assignLoc, elem.get());
+                if (result != UINT32_MAX) {
+                    break;
+                }
+            }
+        },
+        [&](parser::Send *send) { result = heredocPos(ctx, assignLoc, send->receiver.get()); },
+        [&](parser::Node *expr) { result = false; });
 
-bool isAttrAccessorSend(const parser::Send *send) {
-    return (send->receiver == nullptr || parser::isa_node<parser::Self>(send->receiver.get())) &&
-           (send->method == core::Names::attrReader() || send->method == core::Names::attrWriter() ||
-            send->method == core::Names::attrAccessor());
+    return result;
+}
+
+void CommentsAssociator::consumeCommentsBetweenLines(int startLine, int endLine, string kind = "expression") {
```

Comment:
Combined the 2 check dangling comment functions.

Reply:
Let's remove the default for `kind` and always pass a specific node name, this will help us debug if a comment gets reported as unused where we don't expect it.